src/routing_table/mod.rs

```diff
@@ -605,6 +605,40 @@ impl<T: Binary + Clone + Copy + Debug + Default + Hash + Xorable> RoutingTable<T
             .collect()
     }
 
+    /// Returns the `route`-th node in the given group, sorted by distance to `target`
+    pub fn get_routeth_node(&self,
+                            group: &HashSet<T>,
+                            target: T,
+                            exclude: Option<T>,
+                            route: usize)
+                            -> Result<T, Error> {
+        let mut names = if let Some(exclude) = exclude {
+            group.iter().filter(|&x| *x != exclude).collect_vec()
+        } else {
+            group.iter().collect_vec()
+        };
+        names.sort_by(|&lhs, &rhs| target.cmp_distance(lhs, rhs));
+        match names.get(route) {
+            Some(&name) => Ok(*name),
+            None => {
+                // TODO: This is a workaround for the cases where we have not connected to all
+                //       needed contacts yet and may have empty or incomplete groups.
+                // Err(Error::CannotRoute),
+                let cmp = |name0: &T, name1: &T| target.cmp_distance(name0, name1);
+                match self.groups
+                    .iter()
+                    .map(|(prefix, group)| (prefix.lower_bound(), group))
+                    .sorted_by(|&(name0, _), &(name1, _)| cmp(&name0, &name1))
+                    .into_iter()
+                    .flat_map(|(_, peers)| peers.iter().cloned().sorted_by(&cmp).into_iter())
+                    .next() {
```

Comment:
Not sure. But I think in most cases it would make sense. E.g. if route 0 failed in a situation like
```
A0 A1 A2
B0 B1
C0 C1 C2 C3
D0 D1 D2
```
then that was `A0 B0 C0 D0` and it's not clear that `B0` was at fault. So it would still make sense to wrap around and try `A2 B0 C2 D2`.